src/metabase/driver/sql_jdbc/actions.clj

```diff
@@ -303,3 +303,57 @@
             [(conj errors {:index row-index, :error (ex-message e)})
              created-rows]))))
      rows)))
+
+(defn- pk-values->filters [table-id pk-values]
+  ;; check that all pk-values have the same shape
+  (when-not (apply = (map (comp set keys) pk-values))
+    (throw (ex-info (trs "Inconsistent row selection.")
+                    {:status-code 400, :pk-values pk-values})))
+  (let [given-pks (set (keys (first pk-values)))
+        table-pks (into {} (map (juxt :name :id) (filter #(= (:semantic_type %) :type/PK) (:fields (first (table/with-fields [(Table table-id)]))))))]
+    (when-not (= given-pks (set (keys table-pks)))
+      (throw (ex-info (trs "Must select with all PKs.")
+                      {:status-code 400, :given-pks (sort given-pks) :table-pks (sort (keys table-pks))})))
+    (map
+      (fn [pk-value]
+        (into [:and]
+              (mapv (fn [[pk-name value]]
+                      [:= [:field (get table-pks pk-name) nil] value])
+                    pk-value)))
+      pk-values)))
+
```

Comment:
I wanted to write an example that combined the various suggestions I made and I ended up making a few other tweaks:

- Breaking this out into a few smaller functions makes this easier to read, even if the code is a little longer. `pk-value-maps->honeysql-filter-clauses` becomes a simple high-level function that uses a few other functions for validation and for building the filter clauses. This also makes individual pieces testable. You can test the individual validation functions and `pk-value-map->honeysql-filter-clause` in isolation if you don't want to try to write a big e2e test to trigger them. (We definitely need to be writing tests that trigger validation failures)
- `pk-name->id` is a more explicit name for the map of PK Field name -> ID than `table-pks` 
- It's nice to have error messages tell you what they're expecting if you can, or what the valid options are. It saves people from having to figure out how to get that information themselves and having to do so -- it's easier to fix stuff. So I tweaked the error messages to tell you what it expects vs. what you gave it

```clj
(defn- check-consistent-pk-value-map-keys [pk-value-maps]
  (let [all-pk-value-map-column-sets (reduce
                                      (fn [seen-set pk-value-map]
                                        (conj seen-set (set (keys pk-value-map))))
                                      #{}
                                      pk-value-maps)]
    (when (> (count all-pk-value-map-column-sets) 1)
      (throw (ex-info (tru "The specified rows have different sets of columns: {0}"
                           (str/join ", " (map pr-str all-pk-value-map-column-sets)))
                      {:status-code 400, :column-sets all-pk-value-map-column-sets})))))

(defn- check-pk-value-maps-have-expected-columns [pk-value-maps expected-columns]
  ;; we only actually need to check the first map since [[check-consistent-pk-value-map-keys]] should have checked that
  ;; they all have the same keys.
  (let [expected-columns (set expected-columns)
        actual-columns   (set (keys (first pk-value-maps)))]
    (when-not (= actual-columns expected-columns)
      (throw (ex-info (tru "The specified rows have the wrong columns: expected {0}, but got {1}" expected-columns actual-columns)
                      {:status-code 400, :expected-columns expected-columns, :actual-columns actual-columns})))))

(defn- pk-value-map->honeysql-filter-clause [pk-name->id pk-value-map]
  (into [:and] (for [[pk-name value] pk-value-map]
                 [:=
                  [:field (get pk-name->id pk-name) nil]
                  value])))

(defn- pk-value-maps->honeysql-filter-clauses [table-id pk-value-maps]
  ;; fetch a map of PK Field name -> ID for all the PK Fields for the Table
  (let [pk-name->id (db/select-field->id :name Field
                      {:where [:and
                               [:= :table_id table-id]
                               (mdb.u/isa :semantic_type :type/PK)]})]
    ;; validate the keys in `pk-value-maps`
    (check-consistent-pk-value-map-keys)
    (check-pk-value-maps-have-expected-columns pk-value-maps (set (keys pk-name->id)))
    ;; not build the HoneySQL filter clauses
    (mapv (partial pk-value-map->honeysql-filter-clause pk-name->id) pk-value-maps)))
```

I'm not trying to step on your toes and rewrite all the code you've written, but I hope the suggestions I've made and the reasoning for them make sense. Feel free to do with them what you wish. My main goal here is to try

Reply:
Oh yeah one other benefit from splitting these out into separate functions is that we'd be able to tweak them a bit and reuse them for `:bulk/update`